numpy/core/src/umath/ufunc_object.c

```diff
@@ -2802,7 +2836,7 @@ reducelike_promote_and_resolve(PyUFuncObject *ufunc,
      * (although this should possibly happen through a deprecation)
      */
     if (resolve_descriptors(3, ufunc, ufuncimpl,
-            ops, out_descrs, signature, NPY_UNSAFE_CASTING) < 0) {
```

Comment:
So here an below, you are making this a choice that can be made at a higher level?

Reply:
Yeah.  Right now reductions always use unsafe casting.  They still do here, but I thought that `resolve_dtypes` should maybe not.
The downside is that if someone now hardcodes `resolve_dtypes(..., casting="unsafe", reduction=True)` and NumPy eventually changes the `"same_kind"`, they also have to update.

Maybe I should do it `casting=None` and if not given default to "unsafe" (our default) when `reduction=True`?  It did seem OK to support `casting` for `resolve_dtypes` in reductions, which is why this is needed/useful.